pkg/storage/s3/s3.go

```diff
@@ -211,35 +260,69 @@ func (d *driver) CreateStorage(cr *opapi.ImageRegistry, modified *bool) error {
 		return err
 	}
 
-	for i := 0; i < 5000; i++ {
-		if len(d.Config.Bucket) == 0 {
-			d.Config.Bucket = fmt.Sprintf("%s-%s-%s-%s", clusterconfig.StoragePrefix, d.Config.Region, strings.Replace(string(cv.Spec.ClusterID), "-", "", -1), strings.Replace(string(uuid.NewUUID()), "-", "", -1))[0:62]
-		}
-
-		_, err := svc.CreateBucket(&s3.CreateBucketInput{
-			Bucket: aws.String(d.Config.Bucket),
-		})
+	// If a bucket name is supplied, and it already exists and we can access it
+	// just update the config
+	var bucketExists bool
+	if len(d.Config.Bucket) != 0 {
+		bucketExists, err = d.bucketExists(d.Config.Bucket)
 		if err != nil {
 			if aerr, ok := err.(awserr.Error); ok {
 				switch aerr.Code() {
-				case s3.ErrCodeBucketAlreadyExists:
-					if cr.Spec.Storage.S3.Bucket != "" {
-						break
-					}
-					d.Config.Bucket = ""
-					continue
-				default:
+				case s3.ErrCodeNoSuchBucket, "Forbidden", "NotFound":
+					// If the bucket doesn't exist that's ok, we'll try to create it
 					util.UpdateCondition(cr, opapi.StorageExists, operatorapi.ConditionFalse, aerr.Code(), aerr.Error(), modified)
+				default:
+					util.UpdateCondition(cr, opapi.StorageExists, operatorapi.ConditionUnknown, "Unknown Error Occurred", err.Error(), modified)
 					return err
 				}
 			}
 		}
-		break
 	}
+	if len(d.Config.Bucket) != 0 && bucketExists {
+		*cr.Status.Storage.S3 = *d.Config
+		util.UpdateCondition(cr, opapi.StorageExists, operatorapi.ConditionTrue, "S3 Bucket Exists", "User supplied S3 bucket exists and is accessible", modified)
+
+	} else {
+		generatedName := false
+		for i := 0; i < 5000; i++ {
+			// If the bucket name is blank, let's generate one
+			if len(d.Config.Bucket) == 0 {
+				d.Config.Bucket = fmt.Sprintf("%s-%s-%s-%s", clusterconfig.StoragePrefix, d.Config.Region, strings.Replace(string(cv.Spec.ClusterID), "-", "", -1), strings.Replace(string(uuid.NewUUID()), "-", "", -1))[0:62]
+				generatedName = true
+			}
+
+			_, err := svc.CreateBucket(&s3.CreateBucketInput{
+				Bucket: aws.String(d.Config.Bucket),
+			})
+			if err != nil {
+				if aerr, ok := err.(awserr.Error); ok {
+					switch aerr.Code() {
+					case s3.ErrCodeBucketAlreadyExists:
+						if d.Config.Bucket != "" && !generatedName {
+							util.UpdateCondition(cr, opapi.StorageExists, operatorapi.ConditionFalse, "Unable to Access Bucket", "The bucket exists, but we do not have permission to access it", modified)
+							break
+						}
+						d.Config.Bucket = ""
+						continue
+					default:
+						util.UpdateCondition(cr, opapi.StorageExists, operatorapi.ConditionFalse, aerr.Code(), aerr.Error(), modified)
+						return err
+					}
+				}
+			}
+			cr.Status.StorageManaged = true
+			*cr.Status.Storage.S3 = *d.Config
+			*cr.Spec.Storage.S3 = *d.Config
```

Comment:
same problem here. you're shallow copying.

Reply:
updated